### PR TITLE
Add integration tests for elasticsearch #2244

### PR DESCRIPTION
**Pull request by lordlothar99**

### CODE REVIEW

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java [hunk 1]
```diff
@@ -4,8 +4,8 @@
 import <%=packageName%>.<%= mainClass %>;
 import <%=packageName%>.domain.<%= entityClass %>;
 import <%=packageName%>.repository.<%= entityClass %>Repository;<% if (service != 'no') { %>
-import <%=packageName%>.service.<%= entityClass %>Service;<% } else { if (searchEngine == 'elasticsearch') { %>
```
Comment (deepu105) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: why did you remove the else statement?  when there is service layer tests should be triggered via that instead of repository

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java [hunk 2]
```diff
@@ -195,7 +195,8 @@ public void setup() {
 
     @Before
     public void initTest() {<% if (databaseType == 'mongodb' || databaseType == 'cassandra') { %>
-        <%= entityInstance %>Repository.deleteAll();<% } %>
+        <%= entityInstance %>Repository.deleteAll();<% } if (searchEngine == 'elasticsearch') { %>
+        <%= entityInstance %>SearchRepository.deleteAll();<% } %>
```
Comment (deepu105) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: shouldnt this be handled by the ` <%= entityInstance %>Repository.deleteAll();`

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java [hunk 3]
```diff
@@ -231,7 +232,12 @@ public void initTest() {<% if (databaseType == 'mongodb' || databaseType == 'cas
         assertThat(test<%= entityClass %>.is<%=fields[idx].fieldInJavaBeanMethod%>()).isEqualTo(<%='DEFAULT_' + fields[idx].fieldNameUnderscored.toUpperCase()%>);
         <%_ } else { _%>
         assertThat(test<%= entityClass %>.get<%=fields[idx].fieldInJavaBeanMethod%>()).isEqualTo(<%='DEFAULT_' + fields[idx].fieldNameUnderscored.toUpperCase()%>);
-        <%_ }} _%>
+        <%_ }} if (searchEngine == 'elasticsearch') { _%>
+
+        // Validate the <%= entityClass %> in ElasticSearch
+        <%= entityClass %> <%= entityInstance %>Es = <%= entityInstance %>SearchRepository.findOne(test<%= entityClass %>.getId());
```
Comment (deepu105) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: this should be `<%= entityClass %>Service.findOne` when entity has service and else `<%= entityInstance %>SearchRepository.findOne`

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java [hunk 4]
```diff
@@ -308,7 +314,8 @@ public void initTest() {<% if (databaseType == 'mongodb' || databaseType == 'cas
     @Transactional<% } %>
     public void update<%= entityClass %>() throws Exception {
         // Initialize the database
-        <%= entityInstance %>Repository.save<% if (databaseType == 'sql') { %>AndFlush<% } %>(<%= entityInstance %>);
+        <%= entityInstance %>Repository.save<% if (databaseType == 'sql') { %>AndFlush<% } %>(<%= entityInstance %>);<% if (searchEngine == 'elasticsearch') { %>
+        <%= entityInstance %>SearchRepository.save(<%= entityInstance %>);<%_ } _%>
```
Comment (deepu105) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: this is required only if entity doesnt have a service else its handled in the service already

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 2 above)
Comment (deepu105) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: Ok this is needed for non service entity, if entity has service this is handled in its delete method

**Comment by deepu105** [comment]
can you take care of my comments plz. once done ill merge it

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 1 above)
Comment (lordlothar99) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: Direct access to SearchRepository is needed whether service is present or not, cause some new assertions are added in create / update / delete test methods

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 3 above)
Comment (lordlothar99) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: Actually , `EntityService.findOne()` invokes `Entity repository.findOne()`.. While the purpose of this assertion is to read from Elasticsearch. So I need direct usage of `EntitySearchRepository.findOne()`

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 4 above)
Comment (lordlothar99) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: You are right. I just wanted to keep the code similar to the other tests.. But I can change that if you prefer service invocation?

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 2 above)
Comment (lordlothar99) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: yes, we need both `deleteAll()` invocations

**Comment by lordlothar99** [comment]
Hi @deepu105, just let me know what u think about my explanations plz

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 3 above)
Comment (geraldhumphries) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: What about sending a request to `_search/entities` with the query `"id:" + test<%= entityClass %>.getId()`?

Comment (geraldhumphries) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: Never mind, I see that it is following the same pattern as above.

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 2 above)
Comment (geraldhumphries) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: If the mapping of an entity changes, trying to insert an object into an old index can cause exceptions (ex. if a field type changes). Deleting the data is not enough, you need to delete the entity's index then recreate it with the new mapping:
```
        elasticsearchTemplate.deleteIndex(<%= entityClass %>.class);
        elasticsearchTemplate.createIndex(<%= entityClass %>.class);
        elasticsearchTemplate.putMapping(<%= entityClass %>.class);
```

This will also delete the data.

Comment (lordlothar99) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: yeah I read that, actually I was wondering we could just use `deleteAll()`, as it's fast, and during the tests mappings won't change .. no?

Comment (geraldhumphries) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: They won't change during the tests, but the ES indexes persist (they're stored on the hard drive) so the mappings may change between test sessions. That's why I was trying to find a solution to delete all indexes before any test is run.

Comment (deepu105) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: Ill look at them today

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 1 above)
Comment (deepu105) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: ok since you are using `<%= entityInstance %>SearchRepository.deleteAll();` anyway

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 2 above)
Comment (deepu105) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: Ill let you guys decide this

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 3 above)
Comment (deepu105) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: ok

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 4 above)
Comment (deepu105) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: yes since service already does that its a good way to test if service does that so i would prefer to use service when its enabled

**Comment by deepu105** [comment]
Im ok with all except the last save `<%= entityInstance %>SearchRepository.save(<%= entityInstance %>);` stuff I would prefer that via service when its enabled

**Comment by lordlothar99** [comment]
I'm on it

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 2 above)
Comment (lordlothar99) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: @geraldhumphries ok, I understand now that I've read once again your comment. I will try to find a better way to handle that...

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java (hunk 4 above)
Comment (lordlothar99) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: ok I will

**Comment by lordlothar99** [comment]
@deepu105 as we said, I replaced `EntityRepository.save()` and `EntitySearchRepository.save()` by `EntityService.save()` when possible
@geraldhumphries I introduced new `EntityIndexInitializer` beans, which reset ElasticSearch indexes at startup only (`@PostConstruct` method), it takes more or less 2s on my (slow) computer. Still `EntityResourceIntTest` `@Before` method invokes `EntitySearchRepository.deleteAll()` which is very fast (<10ms).
Can you please review these changes, and let me know ?
thx !

**Comment by lordlothar99** [comment]
seems like Travis is failing sometimes, but I think that's nothing to do with this PR, no ?

**Comment by jdubois** [comment]
No we are having other build issues at the moment, don't worry!
Le 18 mars 2016 9:11 AM, "François Lecomte" <notifications@github.com> a
écrit :

File: generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java [hunk 5]
```diff
@@ -361,6 +382,35 @@ public void initTest() {<% if (databaseType == 'mongodb' || databaseType == 'cas
 
         // Validate the database is empty
         List<<%= entityClass %>> <%= entityInstancePlural %> = <%= entityInstance %>Repository.findAll();
-        assertThat(<%= entityInstancePlural %>).hasSize(databaseSizeBeforeDelete - 1);
-    }
+        assertThat(<%= entityInstancePlural %>).hasSize(databaseSizeBeforeDelete - 1);<% if (searchEngine == 'elasticsearch') { %>
+
+        // Validate tElasticSearch is empty
+        boolean <%= entityInstance %>ExistsInEs = <%= entityInstance %>SearchRepository.exists(<%= entityInstance %>.getId());
```
Comment (cicoub13) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: When entityInstance is same that entityInstancePlural, there is a problem here (with entity people for example). 

`<%= entityInstance %>.getId()` does not exist because entityInstance was redefined in a `List<<%= entityClass %>>`

Comment (deepu105) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: @lordlothar99 can you check this plz?

Comment (cicoub13) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: Inversing elasticSearch validation and database validation works.
Replcaing `List<<%= entityClass %>> <%= entityInstancePlural %>` by `List<<%= entityClass %>> <%= entityInstancePlural %>Database` works too.

Comment (lordlothar99) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: I'm on it ; do you prefer a new PR, or update this one ?

Comment (deepu105) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: I think you cant update this as its merged you need to do a new PR

Comment (lordlothar99) [inline generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java]: yeah u r right, I pushed but nothing happened ...
I created fixed #3227
